src/rdkafka_telemetry_decode.c

```diff
@@ -0,0 +1,536 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rd.h"
+#include "rdkafka_telemetry_decode.h"
+#include "nanopb/pb.h"
+#include "nanopb/pb_decode.h"
+#include "opentelemetry/metrics.pb.h"
+#include "rdkafka_int.h"
+#include "rdkafka_telemetry_encode.h"
+#include "rdunittest.h"
+#include "rdkafka_lz4.h"
+#include "rdgz.h"
+#include "rdkafka_zstd.h"
+#include "snappy.h"
+#include "rdfloat.h"
+
+
+#define _NANOPB_STRING_DECODE_MAX_BUFFER_SIZE 1024
+
+struct metric_unit_test_data {
+        rd_kafka_telemetry_metric_type_t type;
+        int32_t current_field;
+        char metric_name[_NANOPB_STRING_DECODE_MAX_BUFFER_SIZE];
+        char metric_description[_NANOPB_STRING_DECODE_MAX_BUFFER_SIZE];
+        int64_t metric_value_int;
+        double metric_value_double;
+        uint64_t metric_time;
+};
+
+static struct metric_unit_test_data unit_test_data;
+
+static void clear_unit_test_data(void) {
+        unit_test_data.type           = RD_KAFKA_TELEMETRY_METRIC_TYPE_GAUGE;
+        unit_test_data.current_field  = 0;
+        unit_test_data.metric_name[0] = '\0';
+        unit_test_data.metric_description[0] = '\0';
+        unit_test_data.metric_value_int      = 0;
+        unit_test_data.metric_time           = 0;
+}
+
+static bool
+decode_string(pb_istream_t *stream, const pb_field_t *field, void **arg) {
+        rd_kafka_telemetry_decode_interface_t *interface      = *arg;
+        uint8_t buffer[_NANOPB_STRING_DECODE_MAX_BUFFER_SIZE] = {0};
+
+        if (stream->bytes_left > sizeof(buffer) - 1) {
+                RD_INTERFACE_CALL(interface, error,
+                                  "String too long for buffer");
+                return false;
+        }
+
+        if (!pb_read(stream, buffer, stream->bytes_left)) {
+                RD_INTERFACE_CALL(interface, error, "Failed to read string");
+                return false;
+        }
+
+        RD_INTERFACE_CALL(interface, decoded_string, buffer);
+        return true;
+}
+
+static bool
+decode_key_value(pb_istream_t *stream, const pb_field_t *field, void **arg) {
+        rd_kafka_telemetry_decode_interface_t *interface = *arg;
+        opentelemetry_proto_common_v1_KeyValue key_value =
+            opentelemetry_proto_common_v1_KeyValue_init_zero;
+        key_value.key.funcs.decode                      = &decode_string;
+        key_value.key.arg                               = interface;
+        key_value.value.value.string_value.funcs.decode = &decode_string;
+        key_value.value.value.string_value.arg          = interface;
+        if (!pb_decode(stream, opentelemetry_proto_common_v1_KeyValue_fields,
+                       &key_value)) {
+                RD_INTERFACE_CALL(interface, error,
+                                  "Failed to decode KeyValue: %s",
+                                  PB_GET_ERROR(stream));
+                return false;
+        }
+        return true;
```

Comment:
Here it can decode tag value if available, for example broker id
```suggestion

        if (key_value.value.which_value ==
            opentelemetry_proto_common_v1_AnyValue_int_value_tag) {
                RD_INTERFACE_CALL(interface, decoded_int64,
                                  key_value.value.value.int_value);
        }

        return true;
```

Reply:
Updated